doc/tutorials/01-lennard_jones/01-lennard_jones.ipynb

```diff
@@ -336,42 +405,142 @@
+    "DAMPING = 30\n",
+    "MAX_STEPS = 10000\n",
+    "MAX_DISPLACEMENT = 0.01 * LJ_SIG\n",
+    "EMSTEP = 10"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "solution2": "hidden",
+    "solution2_first": true
+   },
+   "source": [
+    "**Exercise:**\n",
+    "\n",
+    "* Use <tt>[espressomd.integrate](http://espressomd.org/html/doc/espressomd.html#module-espressomd.minimize_energy).set_steepest_descent</tt> to relax the initial configuration.\n",
+    "  Use a maximal displacement to <tt>MAX_DISPLACEMENT</tt>.\n",
+    "  A damping constant <tt>gamma = DAMPING</tt> usually is a good choice.\n",
+    "* Use the relative change of the system total energy and the maximal force as a convergence criterion.\n",
+    "  See the [<tt>espressomd.analyze</tt>](http://espressomd.org/html/doc/espressomd.html#module-espressomd.analyze) documentation on how to obtain energies and the [<tt>espressomd.particle_data</tt> module](http://espressomd.org/html/doc/espressomd.html#module-espressomd.particle_data) for the forces.\n",
+    "  The steepest descent has converged if the relative force change < <tt>F_TOL</tt>\n",
+    "* Break the minimization loop after a maximal number of <tt>MAX_STEPS</tt> steps or if convergence is achieved.\n",
+    "  Check for convergence every <tt>EMSTEP</tt> steps.\n",
+    "  \n",
+    "***Hint:*** To obtain the initial fores one has to initialize integrator using <tt>integ_steps=0</tt>."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "solution2": "hidden"
+   },
+   "outputs": [],
+   "source": [
+    "# Set up steepest descent integration\n",
+    "system.integrator.set_steepest_descent(f_max=F_MAX,\n",
+    "                                       gamma=DAMPING,\n",
+    "                                       max_displacement=MAX_DISPLACEMENT)\n",
+    "# Initialize integrator to obtain initial forces\n",
+    "system.integrator.run(0)\n",
     "\n",
     "i = 0\n",
-    "act_min_dist = system.analysis.min_dist()\n",
-    "while i < WARM_N_TIME and act_min_dist < MIN_DIST:\n",
-    "    system.integrator.run(WARM_STEPS)\n",
-    "    act_min_dist = system.analysis.min_dist()\n",
-    "    i += 1\n",
-    "    LJ_CAP += 1.0\n",
-    "    system.force_cap = LJ_CAP"
+    "while i < MAX_STEPS//EMSTEP:\n",
+    "    prev_maxforce = maxforce\n",
+    "    prev_energy = system.analysis.energy()['total']\n",
+    "    system.integrator.run(EMSTEP)\n",
+    "    maxforce = np.max(np.abs(system.part[:].f))\n",
+    "    relforce = np.abs((maxforce-prev_maxforce)/prev_maxforce)\n",
+    "    energy = system.analysis.energy()['total']\n",
+    "    relener = np.abs((energy-prev_energy)/prev_energy)\n",
+    "    print(\"minimization step: {:4.0f}\\tmax. rel. force change:{:+3.3e}\\trel. energy change:{:+3.3e}\".format((i+1)*EMSTEP,relforce, relener))\n",
+    "    if relforce < F_TOL or relener < E_TOL:\n",
+    "        break\n",
+    "    i += 1"
```

Comment:
@schlaicha @christophlohrmann should we mention somewhere in your tutorials that this construction is for learning purposes and shouldn't always be copy-pasted in a production script? This construction is not wrong, and the steepest descent integrator was designed to be used in that way (it's referred to as a custom-criterion steepest descent in the user guide), but it won't always work out-of-the-box in a different simulation script:
* the gradient descent stops when the system evolves slowly, which could lead to early exit in dense LJ fluids
* the convergence criterion is based on forces measured by their components while the integrator uses their norm: `np.max(np.abs(system.part[:].f))` vs. `np.max(np.linalg.norm(system.part[:].f, axis=1))`
* the criterion for early exit is based on the maximal force on all particles, whereas the integrator actually works on a subset of all particles

The last point can lead to silent bugs when re-using that steepest descent logic in scripts that set up raspberries or walls of particles. Since fixed particles are skipped by the integrator, their forces won't change, and their forces are typically far larger than any other force in the system, so the derivative of the maximal force is 0. Here is a MWE with a raspberry:
```python
import espressomd
import espressomd.interactions
import numpy as np

system = espressomd.System(box_l=[40.] * 3)
system.time_step = 0.01
system.cell_system.skin = 0.3

system.non_bonded_inter[1, 0].wca.set_params(epsilon=48, sigma=3)
system.non_bonded_inter[1, 1].wca.set_params(epsilon=1, sigma=1)
harmonic_bond = espressomd.interactions.HarmonicBond(k=3000., r_0=3)
system.bonded_inter.add(harmonic_bond)

np.random.seed(0)
center = system.box_l / 2 
system.part.add(id=0, pos=center, type=0, q=-40, fix=3*[1], rotation=3*[1])
for i in range(1, 114):
    colSurfPos = np.random.randn(3)
    colSurfPos = colSurfPos / np.linalg.norm(colSurfPos) * 3 + center
    system.part.add(id=i, pos=colSurfPos, type=1)
    system.part[i].add_bond((harmonic_bond, 0))

F_MAX = 500
system.integrator.set_steepest_descent(f_max=F_MAX, gamma=1.0, max_displacement=0.01)
n = system.integrator.run(10000)
print(f'convergence reached after {n} steps')
print(f'maximal force is {np.max(np.linalg.norm(system.part[:].f, axis=1)):.1f}')
print(f'maximal force excluding fixed particle is {np.max(np.linalg.norm(system.part[1:].f, axis=1)):.1f} < {F_MAX}')
```
output:
```
convergence reached after 46 steps
maximal force is 1989.8
maximal force excluding fixed particle is 464.9 < 500
```

On a side-note, the `F_MAX` variable name looks confusing to me, as it suggests a maximal force when in reality you're using the integrator magic value `0`. I originally thought you had three convergence criteria: maximal force, derivative of the maximal force, derivative of the energy. Maybe `F_MAX_DISABLED = 0` would be more expressive, or using `f_max=0` directly in the parameter list?

Reply:
Sure, I agree that there is *no* generic method to equilibrate your system. However, this statement also applies to the force capping and the steepest descent with a maximal force. I'll add a note for the tutors and maybe one sentence in the notebook.

> the criterion for early exit is based on the maximal force on all particles, whereas the integrator actually works on a subset of all particles

Actually that would be great from a didactic point to mention and maybe include in one of the later tutorials. Is there an easy way to know the subset of particles that the integrator actually uses?

> the convergence criterion is based on forces measured by their components while the integrator uses their norm

Agreed, let's change this in the tutorials.

> using f_max=0 directly in the parameter list

Sounds good to me!

Reply:
Yes, there is no generic method, and force capping is also concerned... I just felt like we should mention it once, otherwise people might copy-paste that block into their simulations. Optimization strategies can be difficult to construct because they depend a lot on the properties of the system. 

\<digression\>While researching optimization strategies and their pitfalls a few years ago, I found a paper that uses an optimization algorithm to find the optimal parameters of another optimization algorithm ([doi:10.1016/0377-2217(94)00350-5](https://doi.org/10.1016/0377-2217(94)00350-5)). It ended up on a slide in one of my group seminars :smiley: \</digression\>

> Actually that would be great from a didactic point to mention and maybe include in one of the later tutorials. Is there an easy way to know the subset of particles that the integrator actually uses?

When particles are fixed in one or more dimensions, the components of the force along the fixed axes are not propagated by integrators. It's not specific to steepest descent. You would have to use `system.part[:].fix` as a mask, e.g. `np.max(np.linalg.norm(system.part[:].f * np.logical_not(system.part[:].fix), axis=1))`, to get the forces propagated by the integrator.

There are also pitfalls with virtual sites, which transfer their force to the tracked particles but only get synchronized at the beginning of each integration step, which causes them to be desynchronized after the last integration step:
```python
import espressomd
import numpy as np
system = espressomd.System(box_l=[40.] * 3)
system.time_step = 0.01
system.cell_system.skin = 0.3
system.non_bonded_inter[1, 1].wca.set_params(epsilon=1, sigma=1)
center = system.box_l / 2 
p1 = system.part.add(id=0, pos=[0, 0, 0], type=1)
p2 = system.part.add(id=1, pos=[0, 0, 0.1], type=1)
p2.vs_auto_relate_to(p1.id)
system.integrator.set_steepest_descent(f_max=800, gamma=1.0, max_displacement=0.01)
n = system.integrator.run(100)
print(f'convergence reached after {n} steps')
print(f'maximal force is {np.max(np.linalg.norm(system.part[:].f, axis=1)):.1f}')
print('system.integrator.run(0)')
system.integrator.run(0)
print(f'maximal force is {np.max(np.linalg.norm(system.part[:].f, axis=1)):.1f}')
```
output:
```
convergence reached after 70 steps
maximal force is 758.7
system.integrator.run(0)
maximal force is 638.0
```
The first value is the one that was propagated by the integrator. The second value represents the true force with the current system state.

Reply:
Should be resolved with 
0ed1854 and 
5104cdf, also see #3891.